sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/UnconditionalJumpStatementBase.cs

```diff
@@ -0,0 +1,147 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class UnconditionalJumpStatementBase<TStatementSyntax, TLanguageKindEnum> : SonarDiagnosticAnalyzer
+        where TStatementSyntax : SyntaxNode
+        where TLanguageKindEnum : struct
+    {
+        protected const string DiagnosticId = "S1751";
+        protected const string MessageFormat = "Remove this '{0}' statement or make it conditional.";

```

Comment:
I am pretty sure we have a rule that should raise here saying that const in generic classes are not shared among instances and that it is recommended to have a non-generic base class (that's why it is done like this for some other rules). I am surprised it didn't raised any issue here.

Note that I am not sure whether it is a big deal in our situation as we know we will have only 2 implementations of this class.

Reply:
Ah, now the base class makes sense. We were wondering with @valhristov why that is. Thanks for clearing that up!

I'll leave it as it is, because as you said, there will only ever be 2 instances and it makes code easier to follow. I think we should follow that pattern (e.g. no extra non-generic base class) in these cases, WDYT?

Reply:
Here is the [RSPEC](https://jira.sonarsource.com/browse/RSPEC-2743) and to make a clarification it impacts `static` fields (probably properties too) so I guess it is all good here.